test/unit/eigensolver/test_band_to_tridiag.cpp

```diff
@@ -43,22 +44,24 @@ const std::vector<std::tuple<SizeType, SizeType, SizeType>> sizes = {
     {4, 6, 3}, {8, 4, 2}, {18, 4, 4}, {34, 6, 6}, {37, 9, 3}  // m != mb
 };
 
-template <class T>
+template <Device D, class T>
 void testBandToTridiag(const blas::Uplo uplo, const SizeType band_size, const SizeType m,
                        const SizeType mb) {
   const LocalElementSize size(m, m);
   const TileElementSize block_size(mb, mb);
 
-  Matrix<T, Device::CPU> mat_a(size, block_size);
-  matrix::util::set_random_hermitian(mat_a);
+  Matrix<T, Device::CPU> mat_a_h(size, block_size);
+  matrix::util::set_random_hermitian(mat_a_h);
 
-  auto [mat_trid, mat_v] = eigensolver::bandToTridiag<Backend::MC>(uplo, band_size, mat_a);
+  MatrixMirror<T, D, Device::CPU> mat_a(mat_a_h);
+  auto [mat_trid, mat_v] = eigensolver::bandToTridiag<Backend::MC>(uplo, band_size, mat_a.get());
+  mat_a.copyTargetToSource();
```

Comment:
Do you like this? IMHO it is more "correct" and readable, but I look forward to your comments.

```suggestion
auto [mat_trid, mat_v] = [&](){
  MatrixMirror<T, D, Device::CPU> mat_a(mat_a_h);
  return eigensolver::bandToTridiag<Backend::MC>(uplo, band_size, mat_a.get());
}();
```

Reply:
What about (if it works) `MatrixMirror<const T, ...`

Reply:
Is the `copyTargetToSource` important in the original? Currently `MatrixMirror<const T, ...>` applies the constness to both target and source, i.e. it will not copy anything in either direction. In theory this could be expanded to have different constness for source and target, and if source is non-const even if target is const, the target could be copied back to the source.

That is to say, `MatrixMirror<const T, ...> should work, but it might not do everything you expect it to do at the moment.

Reply:
I don't see a reason for `copyTargetToSource`... the argument of bandToTridiag is a matrix const T, therefore the matrix cannot be changed...

Reply:
> What about (if it works) `MatrixMirror<const T, ...`

It's a nice idea!


> That is to say, `MatrixMirror<const T, ...> should work, but it might not do everything you expect it to do at the moment.

I was not aware of this feature! I'm not sure I fully understood your point @msimberg (specifically _"i.e. it will not copy anything in either direction"_, probably you meant just with function members). However looking at the code it seems that by having a `MatrixMirror<const T, ...>` we get a nice helper that duplicates, just if needed, a matrix on the target device.

This matrix provided by `MatrixMirror<const T, ...>` will be a `Matrix<const T, TargetDevice>`, and in this case it perfectly fits, because the algorithm requires a const matrix as @rasolca already said.

However, in next PRs I will open today, there is a slightly different use-case, where I need to copy a matrix to the device just if needed, so the helper unifying the code between CPU and GPU versions will be useful, but at the same time I need this as a `Matrix<T>`.

I don't know if it is too much overload for this class and it would be preferable to have a dedicated helper just for this. For the moment it is ok like this, but in the next hours we may have to take a decision also on this side.

Reply:
> I was not aware of this feature! I'm not sure I fully understood your point @msimberg (specifically _"i.e. it will not copy anything in either direction"_, probably you meant just with function members).

Yep, sorry. It will of course copy it to the target on construction. But after that no copies will be made. The non-const version copies back to the source on destruction. But that could also be extended to be controlled by some flag if needed in other situations.

> However, in next PRs I will open today, there is a slightly different use-case, where I need to copy a matrix to the device just if needed, so the helper unifying the code between CPU and GPU versions will be useful, but at the same time I need this as a `Matrix<T>` (EDIT: and I don't need them to be copied back).

Yeah, so this would be a situation where you don't want the automatic copying back. I wasn't entirely sure about making it automatic when I added it, and it might even be better to make it explicit instead. Not sure... As for extending `MatrixMirror` or having another utility, I think the answer will depend on what exactly "if needed" means in your case.

Reply:
With "iff needed" I mean that since the code is shared between CPU and GPU, we have

`MatrixMirror<T, D, CPU>`

that could either become

`MatrixMirror<T, CPU, CPU>` where you don't want to have a copy, it acts as an "alias"

or

`MatrixMirror<T, GPU, CPU>` where you want to allocate on the device and get a clone of the source

Reply:
Yes, the use case you are looking at is the first one, i.e. "the alias" for `MatrixMirror<T, CPU, CPU>`.

But, when the generic code translates to `MatrixMirror<T, GPU, CPU>`, it ends up also copying back on destruction, which is not needed in my next use-case.

This PR is going to be closed soon, I will highlight this problem in my next PR and link to this comment here. 😉